File: Linear Control Course/Chapter20/Lesson1/Chapter20_Lesson1.py
import numpy as np

# Second-order plant parameters (e.g. robot joint)
zeta = 0.3
omega_n = 4.0
k_plant = 1.0  # input gain

def plant_step(x, u):
    """
    x = [position; velocity]
    x_dot = [v; -2 zeta omega_n v - omega_n^2 y + k_plant u]
    """
    y = x[0]
    v = x[1]
    dy = v
    dv = -2.0 * zeta * omega_n * v - (omega_n ** 2) * y + k_plant * u
    return np.array([dy, dv])

# 1-DOF proportional controller
Kp_1dof = 20.0

# 2-DOF controller: same feedback gain, plus first-order prefilter
Ky_2dof = 20.0
Kr_2dof = 20.0
T_f = 0.2  # reference filter time constant

dt = 1e-3
T_end = 2.0
steps = int(T_end / dt)

r = 1.0  # step reference

x_1 = np.zeros(2)  # state for 1-DOF
x_2 = np.zeros(2)  # state for 2-DOF
r_f = 0.0          # filtered reference for 2-DOF

y_hist_1 = np.zeros(steps)
y_hist_2 = np.zeros(steps)

for k_step in range(steps):
    # Current outputs
    y1 = x_1[0]
    y2 = x_2[0]

    # 1-DOF control: u = Kp (r - y)
    u1 = Kp_1dof * (r - y1)

    # 2-DOF control:
    # reference prefilter r_f_dot = (r - r_f) / T_f
    r_f += dt * (r - r_f) / T_f
    # u = Kr r_f - Ky y
    u2 = Kr_2dof * r_f - Ky_2dof * y2

    # Integrate plant dynamics for both loops
    x_1 = x_1 + dt * plant_step(x_1, u1)
    x_2 = x_2 + dt * plant_step(x_2, u2)

    y_hist_1[k_step] = x_1[0]
    y_hist_2[k_step] = x_2[0]

# Compare final values (both should approach 1.0)
print("Final 1-DOF position:", y_hist_1[-1])
print("Final 2-DOF position:", y_hist_2[-1])

# For visualization, plot using matplotlib (not shown here):
# import matplotlib.pyplot as plt
# t = np.arange(steps) * dt
# plt.plot(t, y_hist_1, label="1-DOF")
# plt.plot(t, y_hist_2, label="2-DOF")
# plt.legend(); plt.xlabel("time"); plt.ylabel("y"); plt.show()

File: Linear_Control_Course/Chapter20/Lesson1/Chapter20_Lesson1.m
% Second-order plant parameters
zeta = 0.3;
omega_n = 4.0;
k_plant = 1.0;

Kp1 = 20.0;
Ky2 = 20.0;
Kr2 = 20.0;
Tf = 0.2;

dt = 1e-3;
T_end = 2.0;
steps = round(T_end / dt);

x1_pos = 0.0; x1_vel = 0.0;
x2_pos = 0.0; x2_vel = 0.0;
r = 1.0;
rf = 0.0;

y1 = zeros(steps,1);
y2 = zeros(steps,1);
t = (0:steps-1)' * dt;

for kStep = 1:steps
    y1(kStep) = x1_pos;
    y2(kStep) = x2_pos;

    u1 = Kp1 * (r - x1_pos);

    rf = rf + dt * (r - rf) / Tf;
    u2 = Kr2 * rf - Ky2 * x2_pos;

    dx1_pos = x1_vel;
    dx1_vel = -2*zeta*omega_n*x1_vel ...
              - omega_n^2 * x1_pos + k_plant * u1;
    x1_pos = x1_pos + dt * dx1_pos;
    x1_vel = x1_vel + dt * dx1_vel;

    dx2_pos = x2_vel;
    dx2_vel = -2*zeta*omega_n*x2_vel ...
              - omega_n^2 * x2_pos + k_plant * u2;
    x2_pos = x2_pos + dt * dx2_pos;
    x2_vel = x2_vel + dt * dx2_vel;
end

plot(t, y1, t, y2);
legend('1-DOF','2-DOF');
xlabel('Time (s)');
ylabel('Position');
title('Step response: 1-DOF vs 2-DOF');

% Transfer function representation (Control System Toolbox)
s = tf('s');
G = omega_n^2 / (s^2 + 2*zeta*omega_n*s + omega_n^2);
Cy = Kp1;
F = 1 / (Tf*s + 1);
Cr = Kp1 * F;

T_1dof = feedback(G*Cy, 1);        % Y/R for 1-DOF
T_2dof = (G*Cr) / (1 + G*Cy);      % Y/R for 2-DOF (explicit formula)
